### ISSUE TITLE
Authorization

### ISSUE THREAD

**Issue by bjunc** [nuxt-community/apollo-module#5]
In the times I've used vanilla Apollo, I pass an Authorization header via the network interface `opts`.  This is how the API authorizes graphql access.  I can do that using Nuxt, but I've only gotten as far as hardcoding the token into `nuxt.config.js` in my initial R&D.  For instance:

```
modules: ['@nuxtjs/apollo'],
apollo: {
    clients: {
        default: {
            uri: 'https://example.com/graphql',
            opts: { 
                headers: { 
                    Authorization: 'Bearer ABCDEF1234567'
                }
            }
        }
    }
}
```
So I guess my question is two-fold:

1. Server-side, how can I dynamically set the network interface options?  For instance, setting the Bearer token from a cookie sent in the request.  In a previous version of Nuxt, I was able to do that using a custom auth middleware (`cookie.parse(req.headers['cookie'])`.

2. Once the app hits the browser, if the user's token is stored in local storage, how would I update the network interface with the new options?

**EDIT: See below for current working solution to this**
TL;DR use Nuxt middleware to access Apollo middleware.  Steps: 

1.  Extract the token from an HTTP Only cookie.  You can either:
    1. bake the token extraction logic into an app-specific Apollo middleware
    1. extract in `nuxtServerInit`
1. Assign the token to the store. This makes the token accessible in the browser.  Also necessary if you went the `nuxtServerInit` route, as you need a way to get the token into the next step
1. Create a Nuxt middleware that applies an `Authorization` header via apollo's `applyMiddleware` method.  This is possible because the apollo clients are accessible via `context.app.apolloProvider` (examples below)
1. Run the middleware on any page/layout that apollo auth is needed.

**Comment by bjunc** [comment]
After some more digging on this, I tried a few different options, all of which were dead ends.

I was able to apply a middleware to an Apollo client anywhere `context` is available.  Here is an example within middleware:

```js
let client = context.app.apolloProvider.defaultClient
client.networkInterface.use([{
    applyMiddleware (req, next) {
        if (!req.options.headers) req.options.headers = {}
        req.options.headers['Authorization'] = `Bearer ${jwt}`
        next()
    }
}])
```
> In this example, the JWT is retrieved from a http only cookie.

The problem with this, is that it seems middleware only runs on the server, which is fine for the initial request, but further requests that happen on the browser no longer have the Authorization header set.

I appreciate that this module exists, but I'm not sure how authorization on both server/browser could be handled.  Could the example be expanded to include auth?

**Comment by frankspin89** [comment]
@bjunc Think we can help each other. I only do authorization to the token on the client side at the moment. 

Now trying out a setup where I use the vuex store to set/get the token. Because it's available on both the client/server. You can follow the progress here: https://github.com/nuxt-community/apollo-module/issues/6

**Comment by bjunc** [comment]
Hey @frankspin89 happy to share what I've done so far with this.  Maybe we can help each other. A little context:

Prior to trying the Apollo module (or even the Vue-Apollo plugin), I was able to get this to work with Vanilla Apollo and Nuxt. I used a middleware, which would instantiate an Apollo client (made available throughout the app via setting `context.app.apolloClient` vs this module's convention of `context.app.apolloProvider.defaultClient`).  Server-side, the client would be retrieve a JWT from an HTTP Only cookie.  I called this cookie: `UJSSESSID` (as in Universal JS), which follows similar convention to Symfony's cookie name `PHPSESSID`.  If that cookie didn't exists, the app assumed you were logged out.  Once you've authenticated, the cookie was set.  Since the cookie was also the JWT, this is what was used for any server-side API calls.  This worked well, and basically behaved like a conventional session cookie.

Once the app hits the browser, it needs the token for browser-based API calls.  You _could_ put the token in local storage, but that seems unnecessary to me, as you're essentially storing the token in two places (once as an HTTP Only cookie, and once in local storage).  Rather, since you have the token server-side, simply pass down to the browser app via the initial store state.  The token would be in the store upon app load (assuming you're authenticated).  Then, the same middleware I used before (now `context.isClient`) instantiates an Apollo client using the token from the store (not the token from the cookie).

My attempts to use this module, lead me to want to tweak my Nuxt middleware to apply an Apollo middleware; rather than create a new Apollo instance.  I've gotten that to work, which provides server-side and client-side authentication.

Interested to hear your thoughts on the above.

For the record, I couldn't do this in the `apollo/network-interfaces/default.js` because I need access to the request to grab the cookie server-side, and I need access to the store to grab the token client-side.  Middleware provides both.  So if you want to use this module, you can instantiate a default client, then apply an Apollo middleware (see above).  I also looked into `nuxtServerInit`, but ran into issues with that (I don't recall what the issue was with that). 

Also worth noting, I've seen code snippets referring to `req.session`.  I'm not really sure what this is, maybe something I should have looked into...

**Comment by dohomi** [comment]
Not sure if this is the right approach but it works for me pretty good:
```
authenticateApolloMiddleware.js
export default function ({store, app}) {
  const defaultClient = app.apolloProvider.defaultClient
  defaultClient.networkInterface.use([{
    applyMiddleware (req, next) {
      if (!req.options.headers) {
        req.options.headers = {} // Create the header object if needed.
      }
      const token = store.state && store.state.authToken
      req.options.headers.Authorization = `Bearer ${token}`
      next()
    }
  }])
}

```

**Comment by bjunc** [comment]
Yeah, that's basically where I settled with my implementation.  Only difference is that I didn't use `nuxtServerInit`.  Rather, in my middleware, I check `if (context.isServer)`, and grab from the cookie there.  I think it'd be better to use `nuxtServerInit` if you were making a shareable middleware, that was less opinionated on where the token comes from.  In my case, for my apps, I am sticking with a convention of using `UJSSESSID` for Nuxt session cookies, and so I don't feel the need to extract/set the token externally (who knows, I could change my mind).

Overall though, sounds like we're largely doing the same thing.   Curious if @Atinux has an opinion on this?

**Comment by dohomi** [comment]
@bjunc I have to admit that I haven't done enough testing: the edge starts with following use-case: page-reload on a page where authentication is necessary. The initial request will have the correct credentials because the middleware applied the apollo network middleware with the store value. But now, if the client will do some action (for example search) then the client of $apollo does not have the middleware applied because it ran only on the server
@Atinux is this currently the correct behaviour of the middleware? I thought the middleware is getting called once on server and once on client side before page render?

**Comment by bjunc** [comment]
The above scenario shouldn't be a problem.  Yes, the Apollo middleware is _applied_ on the server, but the Nuxt (and Apollo) middleware will continue to run for subsequent server and browser requests.  Since the token is referencing the store, you can even swap the token in the browser and it'd still work.

It sounds like I should probably post up a repo to show how I have it set up, but here's a work-in-progress snippet from my Nuxt `apollo-auth` middleware:

```js
export default function (context) {

  let token = null
  
  // grab the token from the store's initial state
  if (context.isClient) token = context.store.state.token

  // retrieve the JWT from the cookie
  if (context.isServer) {
    if (!context.req.headers['cookie']) return
    let cookies = cookie.parse(context.req.headers['cookie'])
    if (cookies.UJSSESSID) token = cookies.UJSSESSID
    context.store.state.token = token
  }

  let client = context.app.apolloProvider.defaultClient
  client.networkInterface.use([{
    applyMiddleware (req, next) {
      if (!req.options.headers) req.options.headers = {}
      req.options.headers['Authorization'] = `Bearer ${token}`
      next()
    }
  }])
}
```

As I noted above, this takes an opinionated perspective, and tightly couples the middleware with the session cookie/token logic.  That may not be desirable in your case.  If so, just move the `context.isServer` logic to `nuxtServerInit`

**Comment by dohomi** [comment]
@bjunc thanks for your code snippet. I do basically the same but realized one drawback: do a page reload on a page where u call the middleware. Now do a client side call to your graphql backend ( I do this through a client side search to filter data ). You will realize that the client side search won't have the middleware applied and thus will be without authenticated token. You will have the token applied as soon you do a state change because then the middleware will be called again.
I handle this edge case now like this:
```
AbstractVueClass.vue
  export default {
    created () {
      const networkInterface = this.$apollo.provider.defaultClient.networkInterface
      applyApolloMiddleware(networkInterface, this.$store.state.authToken)
    }
}

authenticateApolloMiddleware.js
import applyApolloMiddleware from '~/util/applyApolloMiddleware'
export default function ({store, app}) {
  const defaultClient = app.apolloProvider.defaultClient
  applyApolloMiddleware(defaultClient.networkInterface, store.state.authToken)
}

utilApplyMiddleware.js
export default function (networkInterface, token) {
  const count = networkInterface._middlewares.length
  if (count) return
  networkInterface.use([{
    applyMiddleware (req, next) {
      if (!req.options.headers) {
        req.options.headers = {} // Create the header object if needed.
      }
      req.options.headers.Authorization = `Bearer ${token}`
      next()
    }
  }])
}


```

**Comment by bjunc** [comment]
That's a good point.  In my prototype, I hadn't actually done a GQL request that wasn't part of a page load.

However, I think a simple solution is to move the logic from a Nuxt middleware to a Nuxt plugin.  Unless I am mistaken, the plugin will run for the server and browser, for every request, which will apply the Apollo middleware in both situations.  I was able to literally copy my entire middleware logic over to a plugin without any modification, and my initial tests worked.  I'll test further later today.

Steps:

1. Copy my code above, and move it to a new plugin file (eg. `/plugins/apollo-auth.js`)
2. Update your `nuxt.config.js`: `plugins: ['~/plugins/apollo-auth']`
3. Delete the `middleware: 'apollo-auth'` you might have had on layouts/pages

`asyncData` will behave the same, but now, you'll be able to make post-load GQL requests with the `Authorization` applied.  Example:

```js
methods: {
    async onSomeButtonClick () {
        let client = this.$apollo.provider.defaultClient
        try {
            await client.query({
                fetchPolicy: 'network-only', // ensure cache isn't being used
                query: gql`{ ... }`
            })
        } catch (err) {
            // error handling
        }
    }
}
```

With this, you won't need the logic you added in your `created` handler.

**Comment by bjunc** [comment]
Here's the plugin's base logic, if you'd rather have the token set outside the plugin (eg. in `nuxtServerInit`).  I left out logic that would bail/throw if `token` and `apolloProvider` didn't exist.

If this works for you, we might want to create a repo for this plugin, and make it a bit more legit.

`~/plugins/apollo-auth.js`
```js
export default (context) => {
   let token = context.store.state.token
   let client = context.app.apolloProvider.defaultClient
   client.networkInterface.use([{
      applyMiddleware (req, next) {
         if (!req.options.headers) req.options.headers = {}
         req.options.headers['Authorization'] = `Bearer ${token}`
         next()
      }
   }])
}
```

**Comment by dohomi** [comment]
@bjunc I just tried your version and moved my code to plugins but then the token is never available on the first SSR request. I guess this happens due to the initial `onServerInit` on the server is running after the first time the plugin is being called. I modified your snippet that it works for me like following:
```
import applyMiddleware from '../util/applyApolloMiddleware'
import { getTokenFromCookie } from '../util/auth-helpers'

export default ({store, app, req, isServer}) => {
  let token
  const client = app.apolloProvider.defaultClient
  if (isServer) {
    token = getTokenFromCookie(req)
  } else {
    token = store.state.authToken
  }
  applyMiddleware(client.networkInterface, token)
}
```

**Comment by bjunc** [comment]
@dohomi My snippet doesn't include the token extraction in an effort to decouple the Apollo auth from the session logic. If you're building the plugin only for yourself, then I don't think it'd hurt to bundle them together (as you've done, and even I did at first).  However, if you break them apart, they would be more modular/flexible.  For instance, if you also wanted to run Axios, and use the token for that as well, you could set the token prior to the plugins running, then have an Axios auth plugin, and an Apollo auth plugin.  To do that, simply move your `getTokenFromCookie` logic to `nuxtServerInit`.  Set the token there, and by the time the `apollo-auth` plugin ran, the token would be set.

**Comment by dohomi** [comment]
@bjunc sounds good. I'm quite happy with the solution now, there should be probably some source on the main README.md about some best practices for new users

**Comment by bjunc** [comment]
@dohomi Yeah, probably so.  I think I'm happy with this approach too.  If I get some time over the weekend, I'll pull together an example repo with a plugin for Apollo auth and session management.  

Out of curiosity, what are you calling your session cookie?  I've been using `UJSSESSID` which follows the `PHPSESSID` convention.

**Comment by dohomi** [comment]
I use this ( I think originally from some example I found ): 
```
export const setToken = ({token, user}) => {
  if (process.server) return
  Cookie.set('access_token', token)
  Cookie.set('user_credentials', JSON.stringify(user))
  window.localStorage.setItem('token', token)
}
```

**Comment by t0ker** [comment]
Would love to see a example :)

**Comment by bjunc** [comment]
@t0ker I'll create an example that uses both a session plugin (for extracting the token from an HTTP Only cookie I call `UJSSESSID`) and an Apollo auth plugin (mostly shown above).  Hopefully over the weekend.

**Comment by dohomi** [comment]
@bjunc I think this issue is resolved. In v2.0 you export a function which holds the whole context as first argument. This is a superset of all our discussions we had earlier

**Comment by bjunc** [comment]
@dohomi v2 of what, Apollo?

I created an initial repo that serves as an example for updating the Apollo client [Nuxt Apollo Auth](https://github.com/bjunc/nuxt-apollo-auth).  Let me know if this is helpful.  I plan on building this out a bit more, as well as creating another one for session logic.  

Somewhat related, I recently started a new project where we are using Nuxt to talk to a Phoenix (Elixir) API.  The authorization basically works as I am showing in the repo above.  Session logic is a combination of using `serverMiddleware`, `nuxtServerInit`, and a small auth middleware.  The combined effect is really pretty awesome, and lightweight.  I didn't have to resort to incorporating Express.  I was thinking of extracting that out into an example repo as well, as I'm not quite happy with the session example on nuxtjs.org

**Comment by bjunc** [comment]
@dohomi Just pulled the `2.0.0` update.  Yep, the fact that context is now sent into network interface pretty much negates any need for the plugin since that's what it did.  Kind of wish @Atinux had chimed in on the conversation regarding that update as to not waste everyone's time trying to figure out a workaround, but oh well.

Here's a quick example for anyone else coming in looking for how to do this in `v2.0.0`:
```js
// ~/apollo/network-interfaces/default.js
// @nuxtjs/apollo v2.0.0

export default ({ store }) => {
  const networkInterface = createNetworkInterface({
    uri: 'http://localhost:4000/graphql',
    transportBatching: true
  })

  let token = store.state.session
  networkInterface.use([{
    applyMiddleware (req, next) {
      if (!req.options.headers) req.options.headers = {}
      req.options.headers['Authorization'] = `Bearer ${token}`
      next()
    }
  }])
  return networkInterface
}
```

This was mostly a copy/paste job out of the plugin into the network interface creation.  Works great alongside the session logic I did for talking to a Phoenix GQL API.

**Comment by dohomi** [comment]
@bjunc to complete your network interface, keep in mind that on client side the store is called after the plugin based on this discussion: https://github.com/nuxt-community/apollo-module/issues/23

This is my solution which works very good. I abondend the use of NUXT.onServerIntit inside of Vuex because the plugin code is triggered before the initialServerInit function.
```javascript
export default ({store, req, isServer}) => {
  let token
  if (isServer) {
    token = getTokenFromCookie(req)
    token && store.commit('SET_AUTH_TOKEN', token)
  } else {
    token = window.__NUXT__.state.authToken
  }
  const networkInterface = createNetworkInterface({
    uri: 'https://api.graph.cool/simple/v1/' + process.env.GRAPH_FILE_API
  })
  applyMiddleware(networkInterface, token)
  return networkInterface
}
```

**Comment by bjunc** [comment]
@dohomi yeah, good call. I realized shortly after I wrote my comment that client calls weren't getting the token set, only server-side calls.  What you did is pretty clever, accessing `__NUXT__.state`, but doesn't that feel a bit dirty?   Almost makes me feel better about doing it the previous plugin way.

Also, not sure if this will help you, but I actually used `serverMiddleware` to handle the token extraction from the cookie.  You need access to the response to be able to set an `HttpOnly` cookie, so it just made sense to keep them coupled (get and set).

```js
// server-middleware/session.js

let cookie = require('cookie')
module.exports = function (req, res, next) {
  
  // req is the Node.js http request object
  // res is the Node.js http response object

  if (req.url === '/logout') {
    res.setHeader('Set-Cookie', cookie.serialize('UJSSESSID', null, { maxAge: 0 }))
  }

  if (req.headers['cookie']) {
    let cookies = cookie.parse(req.headers['cookie'])
    if (cookies.UJSSESSID) req.session = cookies.UJSSESSID
  }

  next()
}
```
> left out the cookie set logic for setting a session, as that would muddy up the example

**Comment by dohomi** [comment]
sure that example would work as well. I just simply wire up the token handling inside of the network plugin directly, because the only piece of code where the token is needed is that exact location where I apply the middleware. I think you can safely close the issue now - we have many ways to deal with vue-apollo thanks again @Atinux

**Comment by kieusonlam** [comment]
Sorry, I have a dummie question. How did you guys set the token into req.session in Nuxt?

**Comment by bjunc** [comment]
@kieusonlam the example above should get you most of the way there.  The JWT is initially assigned to an `HttpOnly` cookie (upon successful authentication), which is extracted and assigned to `req.session` in a server middleware for every server request.  If the cookie doesn't exist, then it's assumed the user is not authenticated.  In your Vuex store`nuxtServerInit`, you have access to `req.session`, which you can then assign to your store state. That will give you a way to know across the whole app whether the user is authenticated, as well as a way to assign the token in other areas that need it for auth headers.  

Worth noting, you can call the cookie anything you want.  I am using `UJSSESSID` per the PHP convention of using `PHPSESSID`.   You could use `MyNuxtToken`, etc..

**Comment by razorabhu1995** [comment]
Did you guys made the repo for this??? I require the step by step process!!

have you guys made it??

**Comment by gomezmark** [comment]
Im a little bit confused.
Did you guys put the other options on a middleware?
